// @syncfusion/ej2-pivotview/src/common/base/css-constant.js
/**
 * CSS Constants
 * @hidden
 */
/** @hidden */
export var ROOT = 'e-pivotfieldlist';
/** @hidden */
export var RTL = 'e-rtl';
/** @hidden */
export var PIVOTCHART_LTR = 'e-ltr';
/** @hidden */
export var DEVICE = 'e-device';
/** @hidden */
export var ICON = 'e-icons';
/** @hidden */
export var ICON_DISABLE = 'e-disable';
/** @hidden */
export var ICON_HIDDEN = 'e-hide';
/** @hidden */
export var AXISFIELD_ICON_CLASS = 'e-dropdown-icon';
/** @hidden */
export var WRAPPER_CLASS = 'e-pivotfieldlist-wrapper';
/** @hidden */
export var OLAP_WRAPPER_CLASS = 'e-olapfieldlist-wrapper';
/** @hidden */
export var CONTAINER_CLASS = 'e-field-list-container';
/** @hidden */
export var TOGGLE_FIELD_LIST_CLASS = 'e-toggle-field-list';
/** @hidden */
export var STATIC_FIELD_LIST_CLASS = 'e-static';
/** @hidden */
export var TOGGLE_SELECT_CLASS = 'e-select-table';
/** @hidden */
export var FIELD_TABLE_CLASS = 'e-field-table';
/** @hidden */
export var BUTTON_DRAGGABLE = 'e-draggable';
/** @hidden */
export var OLAP_FIELD_TABLE_CLASS = 'e-olap-field-table';
/** @hidden */
export var FIELD_LIST_CLASS = 'e-field-list';
/** @hidden */
export var OLAP_FIELD_LIST_CLASS = 'e-olap-field-list-tree';
/** @hidden */
export var FIELD_LIST_TREE_CLASS = 'e-field-list-tree';
/** @hidden */
export var FIELD_HEADER_CLASS = 'e-field-header';
/** @hidden */
export var FIELD_LIST_TITLE_CLASS = 'e-field-list-title';
/** @hidden */
export var FIELD_LIST_TITLE_CONTENT_CLASS = 'e-title-content';
/** @hidden */
export var FIELD_LIST_FOOTER_CLASS = 'e-field-list-footer';
/** @hidden */
export var CALCULATED_FIELD_CLASS = 'e-calculated-field';
/** @hidden */
export var FLAT_CLASS = 'e-flat e-primary';
/** @hidden */
export var OUTLINE_CLASS = 'e-outline';
/** @hidden */
export var AXIS_TABLE_CLASS = 'e-axis-table';
/** @hidden */
export var OLAP_AXIS_TABLE_CLASS = 'e-olap-axis-table';
/** @hidden */
export var LEFT_AXIS_PANEL_CLASS = 'e-left-axis-fields';
/** @hidden */
export var RIGHT_AXIS_PANEL_CLASS = 'e-right-axis-fields';
/** @hidden */
export var ALL_FIELDS_PANEL_CLASS = 'e-all-fields-axis';
/** @hidden */
export var FIELD_PANEL_SCROLL_CLASS = 'e-hscroll';
/** @hidden */
export var AXIS_HEADER_CLASS = 'e-axis-header';
/** @hidden */
export var AXIS_CONTENT_CLASS = 'e-axis-content';
/** @hidden */
export var AXIS_PROMPT_CLASS = 'e-draggable-prompt';
/** @hidden */
export var PIVOT_BUTTON_WRAPPER_CLASS = 'e-pvt-btn-div';
/** @hidden */
export var PIVOT_BUTTON_CLASS = 'e-pivot-button';
/** @hidden */
export var PIVOT_BUTTON_CONTENT_CLASS = 'e-content';
/** @hidden */
export var DRAG_CLONE_CLASS = 'e-button-drag-clone';
/** @hidden */
export var SORT_CLASS = 'e-sort';
/** @hidden */
export var MEMBER_SORT_CLASS = 'e-member-sort';
/** @hidden */
export var SORT_DESCEND_CLASS = 'e-descend';
/** @hidden */
export var FILTER_COMMON_CLASS = 'e-btn-filter';
/** @hidden */
export var FILTER_CLASS = 'e-pv-filter';
/** @hidden */
export var FILTERED_CLASS = 'e-pv-filtered';
/** @hidden */
export var REMOVE_CLASS = 'e-remove';
/** @hidden */
export var DRAG_CLASS = 'e-drag';
/** @hidden */
export var DRAG_DISABLE_CLASS = 'e-disable-drag';
/** @hidden */
export var DROP_INDICATOR_CLASS = 'e-drop-indicator';
/** @hidden */
export var INDICATOR_HOVER_CLASS = 'e-drop-hover';
/** @hidden */
export var MEMBER_EDITOR_DIALOG_CLASS = 'e-member-editor-dialog';
/** @hidden */
export var EDITOR_TREE_WRAPPER_CLASS = 'e-member-editor-wrapper';
/** @hidden */
export var EDITOR_TREE_CONTAINER_CLASS = 'e-member-editor-container';
/** @hidden */
export var DRILLTHROUGH_GRID_CLASS = 'e-drillthrough-grid';
/** @hidden */
export var DRILLTHROUGH_BODY_CLASS = 'e-drillthrough-body';
/** @hidden */
export var DRILLTHROUGH_BODY_HEADER_CONTAINER_CLASS = 'e-drillthrough-body-header-container';
/** @hidden */
export var DRILLTHROUGH_BODY_HEADER_CLASS = 'e-drillthrough-body-header';
/** @hidden */
export var DRILLTHROUGH_BODY_HEADER_COMMON_CLASS = 'e-drillthrough-body-header-common';
/** @hidden */
export var DRILLTHROUGH_BODY_HEADER_VALUE_CLASS = 'e-drillthrough-body-header-value';
/** @hidden */
export var DRILLTHROUGH_DIALOG = 'e-drillthrough-dialog';
/** @hidden */
export var EDITOR_LABEL_WRAPPER_CLASS = 'e-editor-label-wrapper';
/** @hidden */
export var EDITOR_LABEL_CLASS = 'e-editor-label';
/** @hidden */
export var CHECK_BOX_FRAME_CLASS = 'e-frame';
/** @hidden */
export var NODE_CHECK_CLASS = 'e-check';
/** @hidden */
export var NODE_STOP_CLASS = 'e-stop';
/** @hidden */
export var OK_BUTTON_CLASS = 'e-ok-btn';
/** @hidden */
export var CANCEL_BUTTON_CLASS = 'e-cancel-btn';
/** @hidden */
export var ERROR_DIALOG_CLASS = 'e-pivot-error-dialog';
/** @hidden */
export var DROPPABLE_CLASS = 'e-droppable';
/** @hidden */
export var ROW_AXIS_CLASS = 'e-rows';
/** @hidden */
export var COLUMN_AXIS_CLASS = 'e-columns';
/** @hidden */
export var VALUE_AXIS_CLASS = 'e-values';
/** @hidden */
export var FILTER_AXIS_CLASS = 'e-filters';
/** @hidden */
export var GROUPING_BAR_CLASS = 'e-grouping-bar';
/** @hidden */
export var VALUE_COLUMN_CLASS = 'e-value-column';
/** @hidden */
export var GROUP_ALL_FIELDS_CLASS = 'e-group-all-fields';
/** @hidden */
export var GROUP_ROW_CLASS = 'e-group-rows';
/** @hidden */
export var GROUP_COLUMN_CLASS = 'e-group-columns';
/** @hidden */
export var GROUP_FLEX_CLASS = 'e-group-flex';
/** @hidden */
export var GROUP_VALUE_CLASS = 'e-group-values';
/** @hidden */
export var GROUP_FILTER_CLASS = 'e-group-filters';
/** @hidden */
export var DIALOG_CLOSE_ICON_CLASS = 'e-icon-btn';
/** @hidden */
export var NO_DRAG_CLASS = 'e-drag-restrict';
/** @hidden */
export var SELECTED_NODE_CLASS = 'e-list-selected';
/** @hidden */
export var TITLE_HEADER_CLASS = 'e-title-header';
/** @hidden */
export var TITLE_CONTENT_CLASS = 'e-title-content';
/** @hidden */
export var TEXT_CONTENT_CLASS = 'e-text-content';
/** @hidden */
export var FOOTER_CONTENT_CLASS = 'e-footer-content';
/** @hidden */
export var ADAPTIVE_CONTAINER_CLASS = 'e-adaptive-container';
/** @hidden */
export var ADAPTIVE_FIELD_LIST_BUTTON_CLASS = 'e-field-list-btn';
/** @hidden */
export var ADAPTIVE_CALCULATED_FIELD_BUTTON_CLASS = 'e-calculated-field-btn';
/** @hidden */
export var BUTTON_SMALL_CLASS = 'e-small';
/** @hidden */
export var BUTTON_ROUND_CLASS = 'e-round';
/** @hidden */
export var ADD_ICON_CLASS = 'e-add-icon';
/** @hidden */
export var BUTTON_FLAT_CLASS = 'e-flat';
/** @hidden */
export var STATIC_CENTER_DIV_CLASS = 'e-center-div';
/** @hidden */
export var STATIC_CENTER_HEADER_CLASS = 'e-center-title';
/** @hidden */
export var ADAPTIVE_FIELD_LIST_DIALOG_CLASS = 'e-adaptive-field-list-dialog';
/** @hidden */
export var LIST_TEXT_CLASS = 'e-list-text';
/** @hidden */
export var LIST_SELECT_CLASS = 'e-selected-node';
/** @hidden */
export var LIST_FRAME_CLASS = 'e-frame';
/** @hidden */
export var EXCEL_FILTER_ICON_CLASS = 'e-excl-filter-icon';
/** @hidden */
export var SELECTED_MENU_ICON_CLASS = 'e-selected-menu-icon';
/** @hidden */
export var EMPTY_ICON_CLASS = 'e-emptyicon';
/** @hidden */
export var SUB_MENU_CLASS = 'e-submenu';
/** @hidden */
export var FOCUSED_CLASS = 'e-focused';
/** @hidden */
export var SELECTED_CLASS = 'e-selected';
/** @hidden */
export var MENU_ITEM_CLASS = 'e-menu-item';
/** @hidden */
export var FILTER_MENU_OPTIONS_CLASS = 'e-filtermenu-options';
/** @hidden */
export var SELECTED_OPTION_ICON_CLASS = 'e-selected-option-icon';
/** @hidden */
export var SELECTED_LEVEL_ICON_CLASS = 'e-selected-level-icon';
/** @hidden */
export var FILTER_DIV_CONTENT_CLASS = 'e-filter-div-content';
/** @hidden */
export var FILTER_TEXT_DIV_CLASS = 'e-filter-text-div';
/** @hidden */
export var BETWEEN_TEXT_DIV_CLASS = 'e-between-text-div';
/** @hidden */
export var SEPARATOR_DIV_CLASS = 'e-separator-div';
/** @hidden */
export var FILTER_OPTION_WRAPPER_1_CLASS = 'e-filter-option-wrapper-1';
/** @hidden */
export var FILTER_OPTION_WRAPPER_2_CLASS = 'e-filter-option-wrapper-2';
/** @hidden */
export var FILTER_INPUT_DIV_1_CLASS = 'e-filter-input-div-1';
/** @hidden */
export var FILTER_INPUT_DIV_2_CLASS = 'e-filter-input-div-2';
/** @hidden */
export var VALUE_OPTIONS_CLASS = 'e-value-options';
/** @hidden */
export var LEVEL_OPTIONS_CLASS = 'e-level-options';
/** @hidden */
export var FILTER_OPERATOR_CLASS = 'e-filter-operator';
/** @hidden */
export var FILTER_SORT_CLASS = 'e-filter-sort';
/** @hidden */
export var SORT_ASCEND_ICON_CLASS = 'e-sort-ascend-icon';
/** @hidden */
export var SORT_DESCEND_ICON_CLASS = 'e-sort-descend-icon';
/** @hidden */
export var SORT_SELECTED_CLASS = 'e-active';
/** @hidden */
export var COLLAPSE = 'e-collapse';
/** @hidden */
export var EXPAND = 'e-expand';
/** @hidden */
export var TABLE = 'e-table';
/** @hidden */
export var BODY = 'e-body';
/** @hidden */
export var PIVOTBODY = 'e-pivotbody';
/** @hidden */
export var COLUMNSHEADER = 'e-columnsheader';
/** @hidden */
export var ROWSHEADER = 'e-rowsheader';
/** @hidden */
export var VALUESCONTENT = 'e-valuescontent';
/** @hidden */
export var VALUECELL = 'e-valuecell';
/** @hidden */
export var PIVOTHEADER = 'e-pivotheader';
/** @hidden */
export var PGHEADERS = 'e-pgheaders';
/** @hidden */
export var TOPHEADER = 'e-topheader';
/** @hidden */
export var HEADERCELL = 'e-headercell';
/** @hidden */
export var SUMMARY = 'e-summary';
/** @hidden */
export var CELLVALUE = 'e-cellvalue';
/** @hidden */
export var ROW = 'e-row';
/** @hidden */
export var PIVOTTOOLTIP = 'e-pivottooltip';
/** @hidden */
export var TOOLTIP_HEADER = 'e-tooltipheader';
/** @hidden */
export var TOOLTIP_CONTENT = 'e-tooltipcontent';
/** @hidden */
export var NEXTSPAN = 'e-nextspan';
/** @hidden */
export var LASTSPAN = 'e-lastspan';
/** @hidden */
export var EDITOR_SEARCH_WRAPPER_CLASS = 'e-editor-search-wrapper';
/** @hidden */
export var EDITOR_SEARCH_CLASS = 'e-editor-search';
/* eslint-disable */
/** @hidden */
export var EDITOR_SEARCH__INPUT_CLASS = 'e-editor-search-input';
/* eslint-enable */
/** @hidden */
export var SELECT_ALL_WRAPPER_CLASS = 'e-select-all-wrapper';
/** @hidden */
export var SELECT_ALL_CLASS = 'e-select-all';
/** @hidden */
export var PIVOTCALC = 'e-pivot-calc';
/** @hidden */
export var CALCDIALOG = 'e-pivot-calc-dialog-div';
/** @hidden */
export var OLAP_CALCDIALOG = 'e-olap-calc-dialog-div';
/** @hidden */
export var CALCRADIO = 'e-pivot-calc-radio';
/** @hidden */
export var CALCCHECK = 'e-pivot-calc-check';
/** @hidden */
export var CALCINPUT = 'e-pivot-calc-input';
/** @hidden */
export var CALC_FORMAT_INPUT = 'e-custom-format-input';
/** @hidden */
export var CALCINPUTDIV = 'e-pivot-calc-input-div';
/** @hidden */
export var CALC_CUSTOM_FORMAT_INPUTDIV = 'e-pivot-calc-custom-format-div';
/** @hidden */
export var OLAP_CALC_CUSTOM_FORMAT_INPUTDIV = 'e-olap-calc-custom-format-div';
/** @hidden */
export var CALC_HIERARCHY_LIST_DIV = 'e-olap-calc-hierarchy-list-div';
/** @hidden */
export var CALC_FORMAT_TYPE_DIV = 'e-olap-calc-format-type-div';
/** @hidden */
export var CALC_MEMBER_TYPE_DIV = 'e-olap-calc-member-type-div';
/** @hidden */
export var MEMBER_OPTIONS_CLASS = 'e-member-options';
/** @hidden */
export var HIERARCHY_OPTIONS_CLASS = 'e-hierarchy-options';
/** @hidden */
export var FORMAT_OPTIONS_CLASS = 'e-format-options';
/** @hidden */
export var FORMAT_INPUT_CLASS = 'e-format-input-options';
/** @hidden */
export var CALCOUTERDIV = 'e-pivot-calc-outer-div';
/** @hidden */
export var OLAP_CALCOUTERDIV = 'e-olap-calc-outer-div';
/** @hidden */
export var FLAT = 'e-flat';
/** @hidden */
export var FORMAT = 'e-format';
/** @hidden */
export var FORMULA = 'e-pivot-formula';
/** @hidden */
export var TREEVIEW = 'e-pivot-treeview';
/** @hidden */
export var TREEVIEWOUTER = 'e-pivot-treeview-outer';
/** @hidden */
export var CALCCANCELBTN = 'e-pivot-cancel-button';
/** @hidden */
export var CALCADDBTN = 'e-pivot-add-button';
/** @hidden */
export var CALCOKBTN = 'e-pivot-ok-button';
/** @hidden */
export var CALCACCORD = 'e-pivot-accord';
/** @hidden */
export var CALCBUTTONDIV = 'e-pivot-button-div';
/** @hidden */
export var AXIS_ICON_CLASS = 'e-axis';
/** @hidden */
export var AXIS_ROW_CLASS = 'e-axis-row';
/** @hidden */
export var AXIS_COLUMN_CLASS = 'e-axis-column';
/** @hidden */
export var AXIS_VALUE_CLASS = 'e-axis-value';
/** @hidden */
export var AXIS_FILTER_CLASS = 'e-axis-filter';
/** @hidden */
export var AXIS_NAVIGATE_WRAPPER_CLASS = 'e-axis-nav-div';
/** @hidden */
export var LEFT_NAVIGATE_WRAPPER_CLASS = 'e-left-nav-wrapper';
/** @hidden */
export var RIGHT_NAVIGATE_WRAPPER_CLASS = 'e-right-nav-wrapper';
/** @hidden */
export var LEFT_NAVIGATE_CLASS = 'e-left-nav-icon';
/** @hidden */
export var RIGHT_NAVIGATE_CLASS = 'e-right-nav-icon';
/** @hidden */
export var GRID_CLASS = 'e-grid';
/** @hidden */
export var PIVOT_VIEW_CLASS = 'e-pivotview';
/** @hidden */
export var PIVOT_ALL_FIELD_TITLE_CLASS = 'e-pivot-all-field-title';
/** @hidden */
export var PIVOT_FIELD_TITLE_CLASS = 'e-pivot-field-name-title';
/** @hidden */
export var PIVOT_FORMULA_TITLE_CLASS = 'e-pivot-formula-title';
/** @hidden */
export var OLAP_HIERARCHY_TITLE_CLASS = 'e-olap-hierarchy-title';
/** @hidden */
export var OLAP_FORMAT_TITLE_CLASS = 'e-olap-format-title';
/** @hidden */
export var OLAP_MEMBER_TITLE_CLASS = 'e-olap-member-title';
/** @hidden */
export var PIVOT_CONTEXT_MENU_CLASS = 'e-pivot-context-menu';
/** @hidden */
export var MENU_DISABLE = 'e-disabled';
/** @hidden */
export var MENU_HIDE = 'e-menu-hide';
/** @hidden */
export var EMPTY_MEMBER_CLASS = 'e-member-prompt';
/** @hidden */
export var CALC_EDIT = 'e-edit';
/** @hidden */
export var CALC_EDITED = 'e-edited';
/** @hidden */
export var CALC_INFO = 'e-info';
/** @hidden */
export var EMPTY_FIELD = 'e-empty-field';
/** @hidden */
export var FORMAT_DIALOG = 'e-pivot-formatting-dialog';
/** @hidden */
export var FORMAT_CONDITION_BUTTON = 'e-format-condition-button';
/** @hidden */
export var FORMAT_NEW = 'e-new-format';
/** @hidden */
export var FORMAT_OUTER = 'e-format-outer-div';
/** @hidden */
export var FORMAT_INNER = 'e-format-inner-div';
/** @hidden */
export var FORMAT_TABLE = 'e-format-table';
/** @hidden */
export var FORMAT_VALUE_LABEL = 'e-format-value-label';
/** @hidden */
export var FORMAT_LABEL = 'e-format-label';
/** @hidden */
export var INPUT = 'e-input';
/** @hidden */
export var FORMAT_VALUE1 = 'e-format-value1';
/** @hidden */
export var FORMAT_VALUE2 = 'e-format-value2';
/** @hidden */
export var FORMAT_VALUE_SPAN = 'e-format-value-span';
/** @hidden */
export var FORMAT_FONT_COLOR = 'e-format-font-color';
/** @hidden */
export var FORMAT_BACK_COLOR = 'e-format-back-color';
/** @hidden */
export var FORMAT_VALUE_PREVIEW = 'e-format-value-preview';
/** @hidden */
export var FORMAT_COLOR_PICKER = 'e-format-color-picker';
/** @hidden */
export var FORMAT_DELETE_ICON = 'e-format-delete-icon';
/** @hidden */
export var FORMAT_DELETE_BUTTON = 'e-format-delete-button';
/** @hidden */
export var SELECTED_COLOR = 'e-selected-color';
/** @hidden */
export var DIALOG_HEADER = 'e-dlg-header';
/** @hidden */
export var FORMAT_APPLY_BUTTON = 'e-format-apply-button';
/** @hidden */
export var FORMAT_CANCEL_BUTTON = 'e-format-cancel-button';
/** @hidden */
export var FORMAT_ROUND_BUTTON = 'e-small e-round';
/** @hidden */
export var VIRTUALTRACK_DIV = 'e-virtualtrack';
/** @hidden */
export var MOVABLECONTENT_DIV = 'e-movablecontent';
/** @hidden */
export var MOVABLESCROLL_DIV = 'e-movablescrollbar';
/** @hidden */
export var MOVABLECHILD_DIV = 'e-movablechild';
/** @hidden */
export var FROZENCONTENT_DIV = 'e-frozencontent';
/** @hidden */
export var MOVABLEHEADER_DIV = 'e-movableheader';
/** @hidden */
export var FROZENHEADER_DIV = 'e-frozenheader';
/** @hidden */
export var DEFER_APPLY_BUTTON = 'e-defer-apply-button';
/** @hidden */
export var DEFER_CANCEL_BUTTON = 'e-defer-cancel-button';
/** @hidden */
export var LAYOUT_FOOTER = 'e-layout-footer';
/** @hidden */
export var CELL_SELECTED_BGCOLOR = 'e-cellselectionbackground';
/** @hidden */
export var SELECTED_BGCOLOR = 'e-selectionbackground';
/** @hidden */
export var BUTTON_LAYOUT = 'e-button-layout';
/** @hidden */
export var CHECKBOX_LAYOUT = 'e-checkbox-layout';
/** @hidden */
export var DEFER_UPDATE_BUTTON = 'e-defer-update-btn';
/** @hidden */
export var HEADERCONTENT = 'e-headercontent';
/** @hidden */
export var BACK_ICON = 'e-field-list-back-icon';
/** @hidden */
export var TITLE_MOBILE_HEADER = 'e-title-mobile-header';
/** @hidden */
export var TITLE_MOBILE_CONTENT = 'e-title-mobile-content';
/** @hidden */
export var ROW_CELL_CLASS = 'e-rowcell';
/** @hidden */
export var CELL_ACTIVE_BGCOLOR = 'e-active';
/** @hidden */
export var SPAN_CLICKED = 'e-spanclicked';
/** @hidden */
export var ROW_SELECT = 'e-rowselect';
/** @hidden */
export var GRID_HEADER = 'e-gridheader';
/** @hidden */
export var GRID_CONTENT = 'e-gridcontent';
/** @hidden */
export var GRID_EXPORT = 'e-export';
/** @hidden */
export var PIVOTVIEW_EXPORT = 'e-pivotview-export';
/** @hidden */
export var PIVOTVIEW_GRID = 'e-pivotview-grid';
/** @hidden */
export var PIVOTVIEW_EXPAND = 'e-pivotview-expand';
/** @hidden */
export var PIVOTVIEW_COLLAPSE = 'e-pivotview-collapse';
/** @hidden */
export var PIVOTVIEW_GROUP = 'e-pivotview-group';
/** @hidden */
export var PIVOTVIEW_UN_GROUP = 'e-pivotview-ungroup';
/** @hidden */
export var GRID_PDF_EXPORT = 'e-pivotview-pdf-export';
/** @hidden */
export var GRID_EXCEL_EXPORT = 'e-pivotview-excel-export';
/** @hidden */
export var GRID_CSV_EXPORT = 'e-pivotview-csv-export';
/** @hidden */
export var GRID_PNG_EXPORT = 'e-pivotview-png-export';
/** @hidden */
export var GRID_JPEG_EXPORT = 'e-pivotview-jpeg-export';
/** @hidden */
export var GRID_SVG_EXPORT = 'e-pivotview-svg-export';
/** @hidden */
export var GRID_LOAD = 'e-load-report';
/** @hidden */
export var GRID_NEW = 'e-new-report';
/** @hidden */
export var GRID_RENAME = 'e-rename-report';
/** @hidden */
export var GRID_REMOVE = 'e-remove-report';
/** @hidden */
export var GRID_SAVEAS = 'e-saveas-report';
/** @hidden */
export var GRID_SAVE = 'e-save-report';
/** @hidden */
export var GRID_SUB_TOTAL = 'e-sub-total';
/** @hidden */
export var GRID_GRAND_TOTAL = 'e-grand-total';
/** @hidden */
export var GRID_FORMATTING = 'e-toolbar-formatting';
/** @hidden */
export var GRID_TOOLBAR = 'e-pivot-toolbar';
/** @hidden */
export var GRID_REPORT_LABEL = 'e-pivotview-report-label';
/** @hidden */
export var GRID_REPORT_INPUT = 'e-pivotview-report-input';
/** @hidden */
export var GRID_REPORT_OUTER = 'e-report-outer';
/** @hidden */
export var GRID_REPORT_DIALOG = 'e-pivotview-report-dialog';
/** @hidden */
export var TOOLBAR_FIELDLIST = 'e-toolbar-fieldlist';
/** @hidden */
export var TOOLBAR_GRID = 'e-toolbar-grid';
/** @hidden */
export var TOOLBAR_CHART = 'e-toolbar-chart';
/** @hidden */
export var REPORT_LIST_DROP = 'e-reportlist-drop';
/** @hidden */
export var PIVOTCHART = 'e-pivotchart';
/** @hidden */
export var GROUP_CHART_ROW = 'e-group-chart-rows';
/** @hidden */
export var GROUP_CHART_COLUMN = 'e-group-chart-columns';
/** @hidden */
export var GROUP_CHART_VALUE = 'e-group-chart-values';
/** @hidden */
export var GROUP_CHART_MULTI_VALUE = 'e-group-chart-multi-values';
/** @hidden */
export var GROUP_CHART_ACCUMULATION_COLUMN = 'e-group-chart-accumulation-column';
/** @hidden */
export var GROUP_CHART_FILTER = 'e-group-chart-filters';
/** @hidden */
export var GROUP_CHART_VALUE_DROPDOWN_DIV = 'e-group-chart-values-drodown-div';
/** @hidden */
export var GROUP_CHART_VALUE_DROPDOWN = 'e-group-chart-values-drodown';
/** @hidden */
export var GROUP_CHART_COLUMN_DROPDOWN_DIV = 'e-group-chart-columns-dropdown-div';
/** @hidden */
export var GROUP_CHART_COLUMN_DROPDOWN = 'e-group-chart-columns-dropdown';
/** @hidden */
export var CHART_GROUPING_BAR_CLASS = 'e-chart-grouping-bar';
/** @hidden */
export var PIVOT_DISABLE_ICON = 'e-pivotview-disable-icon';
/** @hidden */
export var PIVOT_SELECT_ICON = 'e-pivotview-select-icon';
/** @hidden */
export var VALUESHEADER = 'e-valuesheader';
/** @hidden */
export var ICON_ASC = 'e-icon-ascending';
/** @hidden */
export var ICON_DESC = 'e-icon-descending';
/** @hidden */
export var GRID_GROUPING_BAR_CLASS = 'e-pivot-grouping-bar';
/** @hidden */
export var MDX_QUERY = 'e-mdx-query';
/** @hidden */
export var MDX_QUERY_CONTENT = 'e-mdx-query-content';
/** @hidden */
export var GRID_MDX_DIALOG = 'e-pivotview-mdx-dialog';
/** @hidden */
export var GRID_MDX = 'e-mdx';
/** @hidden */
export var FORMATTING_DIALOG = 'e-pivot-format-dialog';
/** @hidden */
export var FORMATTING_DIALOG_OUTER = 'e-pivot-format-dialog-outer';
/** @hidden */
export var FORMATTING_VALUE_LABLE = 'e-pivot-format-value-lable';
/** @hidden */
export var FORMATTING_VALUE_DROP = 'e-pivot-format-value-drop';
/** @hidden */
export var FORMATTING_FORMAT_LABLE = 'e-pivot-format-lable';
/** @hidden */
export var FORMATTING_FORMAT_DROP = 'e-pivot-format-drop';
/** @hidden */
export var FORMATTING_CUSTOM_LABLE = 'e-pivot-format-custom-lable';
/** @hidden */
export var FORMATTING_CUSTOM_TEXT = 'e-pivot-format-custom-text';
/** @hidden */
export var FORMATTING_SYMBOL_LABLE = 'e-pivot-format-symbol-lable';
/** @hidden */
export var FORMATTING_SYMBOL_DROP = 'e-pivot-format-symbol-drop';
/** @hidden */
export var FORMATTING_GROUPING_LABLE = 'e-pivot-format-grouping-lable';
/** @hidden */
export var FORMATTING_GROUPING_DROP = 'e-pivot-format-grouping-drop';
/** @hidden */
export var FORMATTING_DECIMAL_LABLE = 'e-pivot-format-decimal-lable';
/** @hidden */
export var FORMATTING_DECIMAL_DROP = 'e-pivot-format-decimal-drop';
/** @hidden */
export var FORMATTING_TOOLBAR = 'e-pivot-format-toolbar';
/** @hidden */
export var FORMATTING_TABLE = 'e-pivot-format-table';
/** @hidden */
export var FORMATTING_MENU = 'e-pivot-format-menu';
/** @hidden */
export var NUMBER_FORMATTING_MENU = 'e-pivot-number-format-menu';
/** @hidden */
export var EMPTY_FORMAT = 'e-pivot-conditional-empty-format';
/** @hidden */
export var CONDITIONAL_FORMATTING_MENU = 'e-pivot-conditional-format-menu';
/** @hidden */
export var PIVOTCHART_INNER = 'e-pivotchart-inner';
/** @hidden */
export var PIVOTCHART_TYPE_DIALOG = 'e-pivotchart-type-dialog';
/** @hidden */
export var FORMAT_FONT_COLOR_PICKER = 'e-format-font-color-picker';
/** @hidden */
export var GROUP_PIVOT_ROW = 'e-group-pivot-rows';
