// ng-sortgrid/esm2015/lib/sort/reflection/ngsg-reflect.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sort/reflection/ngsg-reflect.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { NgsgElementsHelper } from '../../helpers/element/ngsg-elements.helper';
import { NgsgStoreService } from '../../store/ngsg-store.service';
import * as i0 from "@angular/core";
import * as i1 from "../../store/ngsg-store.service";
export class NgsgReflectService {
    /**
     * @param {?} ngsgStore
     */
    constructor(ngsgStore) {
        this.ngsgStore = ngsgStore;
    }
    /**
     * @param {?} key
     * @param {?} element
     * @return {?}
     */
    reflectChanges(key, element) {
        /** @type {?} */
        const items = this.ngsgStore.getItems(key);
        /** @type {?} */
        const selectedElements = this.ngsgStore.getSelectedItems(key);
        /** @type {?} */
        const selectedElementIndices = this.getSelectedElementsIndices(selectedElements);
        /** @type {?} */
        const selectedItems = this.getSelectedItems(items, selectedElementIndices);
        /** @type {?} */
        const sortedIndices = selectedElementIndices.sort();
        /** @type {?} */
        const dropIndex = this.findDropIndex(selectedElements, element);
        while (sortedIndices.length > 0) {
            items.splice(sortedIndices.pop(), 1);
        }
        /** @type {?} */
        const result = this.getReflectedItems(items, selectedItems, dropIndex);
        this.ngsgStore.setItems(key, result);
        return result;
    }
    /**
     * @private
     * @param {?} items
     * @param {?} selectedItems
     * @param {?} dropIndex
     * @return {?}
     */
    getReflectedItems(items, selectedItems, dropIndex) {
        /** @type {?} */
        const beforeSelection = items.slice(0, dropIndex);
        /** @type {?} */
        const afterSelection = items.slice(dropIndex, items.length);
        return [...beforeSelection, ...selectedItems, ...afterSelection];
    }
    /**
     * @private
     * @param {?} items
     * @param {?} selectedElementIndexes
     * @return {?}
     */
    getSelectedItems(items, selectedElementIndexes) {
        /** @type {?} */
        const selectedItems = [];
        selectedElementIndexes.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => {
            selectedItems.push(items[index]);
        }));
        return selectedItems;
    }
    /**
     * @private
     * @param {?} selectedElements
     * @return {?}
     */
    getSelectedElementsIndices(selectedElements) {
        return selectedElements.map((/**
         * @param {?} selectedElement
         * @return {?}
         */
        (selectedElement) => selectedElement.originalIndex));
    }
    /**
     * @private
     * @param {?} selectedElements
     * @param {?} element
     * @return {?}
     */
    findDropIndex(selectedElements, element) {
        if (this.isDropInSelection(selectedElements, element)) {
            return NgsgElementsHelper.findIndex(selectedElements[0].node);
        }
        return NgsgElementsHelper.findIndex(element);
    }
    /**
     * @private
     * @param {?} collection
     * @param {?} dropElement
     * @return {?}
     */
    isDropInSelection(collection, dropElement) {
        return !!collection.find((/**
         * @param {?} dragElment
         * @return {?}
         */
        (dragElment) => dragElment.node === dropElement));
    }
}
NgsgReflectService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgsgReflectService.ctorParameters = () => [
    { type: NgsgStoreService }
];
/** @nocollapse */ NgsgReflectService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgsgReflectService_Factory() { return new NgsgReflectService(i0.ɵɵinject(i1.NgsgStoreService)); }, token: NgsgReflectService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgsgReflectService.prototype.ngsgStore;
}
