// @syncfusion/ej2-charts/helpers/e2e/chart-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const e2e_1 = require("@syncfusion/ej2-base/helpers/e2e");
class ChartHelper extends e2e_1.TestHelper {
    constructor(id, wrapperFn) {
        super();
        this.id = id;
        if (wrapperFn !== undefined) {
            this.wrapperFn = wrapperFn;
        }
        return this;
    }
    getChartContainer() {
        return this.selector('#' + this.id);
    }
    getAxisInsideElement() {
        return this.selector('#' + this.id + 'AxisInsideCollection');
    }
    getAxisOutsideElement() {
        return this.selector('#' + this.id + 'AxisOutsideCollection');
    }
    getSeriesElement() {
        return this.selector('#' + this.id + 'SeriesCollection');
    }
    getTooltipElement() {
        return this.selector('#' + this.id + '_tooltip');
    }
    getLegendElement() {
        return this.selector('#' + this.id + '_chart_legend_collections');
    }
    getAnnotationElement() {
        return this.selector('#' + this.id + '_Annotation_Collections');
    }
    getUserInteractionElement() {
        return this.selector('#' + this.id + '_UserInteraction');
    }
    getTrendLineElement() {
        return this.selector('#' + this.id + 'TrendLineCollection');
    }
    getIndicatorElement() {
        return this.selector('#' + this.id + 'IndicatorCollection');
    }
    getZoomingKitElement() {
        return this.selector('#' + this.id + '_Zooming_KitCollection');
    }
    getStriplineBehindCollection() {
        return this.selector('#' + this.id + '_stripline_Behind_collections');
    }
    getStriplineOverCollection() {
        return this.selector('#' + this.id + '_stripline_Over_collections');
    }
}
exports.ChartHelper = ChartHelper;
