// @syncfusion/ej2-charts/helpers/e2e/stock-chart-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const e2e_1 = require("@syncfusion/ej2-base/helpers/e2e");
class StockChartHelper extends e2e_1.TestHelper {
    constructor(id, wrapperFn) {
        super();
        this.id = id;
        if (wrapperFn !== undefined) {
            this.wrapperFn = wrapperFn;
        }
        return this;
    }
    getStockContainer() {
        return this.selector('#' + this.id);
    }
    getStockBorder() {
        return this.selector('#' + this.id + '_stock_border');
    }
    getStockSecondaryElement() {
        return this.selector('#' + this.id + '_Secondary_Element');
    }
    getStockTitle() {
        return this.selector('#' + this.id + '_stockChart_Title');
    }
    getStockRangeLeftTooltip() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_leftTooltip');
    }
    getStockRangeRightTooltip() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_rightTooltip');
    }
    getStockToolbar() {
        return this.selector('#' + this.id + '_Secondary_Element_Secondary_Element');
    }
    getStockTooltip() {
        return this.selector('#' + this.id + '_stockChart_chart_tooltip');
    }
    getStockSvgElement() {
        return this.selector('#' + this.id + '_stockChart_svg');
    }
    getStockChartElement() {
        return this.selector('#' + this.id + '_stockChart_chart');
    }
    getStockChartBorder() {
        return this.selector('#' + this.id + '_stockChart_chart_ChartBorder');
    }
    getStockChartAxisInsideCollection() {
        return this.selector('#' + this.id + '_stockChart_chartAxisInsideCollection');
    }
    getStockChartAxisOutsideCollection() {
        return this.selector('#' + this.id + '_stockChart_chartAxisOutsideCollection');
    }
    getStockChartSeriesCollection() {
        return this.selector('#' + this.id + '_stockChart_chart');
    }
    getStockChartUserInteraction() {
        return this.selector('#' + this.id + '_stockChart_chart_UserInteraction');
    }
    getStockRangeElement() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector');
    }
    getStockRangeBorder() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_ChartBorder');
    }
    getStockRangeGridLines() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_GridLines');
    }
    getStockRangeAxisLabels() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_AxisLabels');
    }
    getStockRangeChart() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_chart');
    }
    getStockRangeSliders() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_sliders');
    }
    getStockRangeLeftUnselectedArea() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_leftUnSelectedArea');
    }
    getStockRangeRightUnselectedArea() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_rightUnSelectedArea');
    }
    getStockRangeSelectedArea() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_SelectedArea');
    }
    getStockRangeLeftSlider() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_LeftSlider');
    }
    getStockRangeRightSlider() {
        return this.selector('#' + this.id + '_stockChart_rangeSelector_RightSlider');
    }
}
exports.StockChartHelper = StockChartHelper;
