// ng-sortgrid/esm2015/lib/helpers/scroll/scroll-helper.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/helpers/scroll/scroll-helper.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
/**
 * @record
 */
export function ScrollPoints() { }
if (false) {
    /** @type {?|undefined} */
    ScrollPoints.prototype.top;
    /** @type {?|undefined} */
    ScrollPoints.prototype.bottom;
}
export class ScrollHelperService {
    /**
     * @param {?} document
     */
    constructor(document) {
        this.document = document;
        this.DEFAULT_SCROLLSPEED = 50;
        this.SCROLL_BUFFER = 50;
        this.window = document.defaultView;
    }
    /**
     * @param {?} event
     * @param {?=} scrollPoints
     * @param {?=} scrollSpeed
     * @return {?}
     */
    scrollIfNecessary(event, scrollPoints = {}, scrollSpeed) {
        /** @type {?} */
        const currentPosition = event.pageY - this.window.scrollY;
        if (this.isTopScrollNeeded(currentPosition, scrollPoints.top)) {
            this.window.scrollBy({ top: -scrollSpeed || -this.DEFAULT_SCROLLSPEED, behavior: 'smooth' });
            return;
        }
        if (this.isBottomScrollNeeded(currentPosition, scrollPoints.bottom)) {
            this.window.scrollBy({ top: scrollSpeed || this.DEFAULT_SCROLLSPEED, behavior: 'smooth' });
        }
    }
    /**
     * @private
     * @param {?} currentPosition
     * @param {?} scrollPointTop
     * @return {?}
     */
    isTopScrollNeeded(currentPosition, scrollPointTop) {
        return scrollPointTop ? currentPosition < scrollPointTop :
            currentPosition < this.SCROLL_BUFFER;
    }
    /**
     * @private
     * @param {?} currentPosition
     * @param {?} scrollPointBottom
     * @return {?}
     */
    isBottomScrollNeeded(currentPosition, scrollPointBottom) {
        return scrollPointBottom ? currentPosition > scrollPointBottom :
            currentPosition > this.window.innerHeight - this.SCROLL_BUFFER;
    }
}
ScrollHelperService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ScrollHelperService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
/** @nocollapse */ ScrollHelperService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ScrollHelperService_Factory() { return new ScrollHelperService(i0.ɵɵinject(i1.DOCUMENT)); }, token: ScrollHelperService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ScrollHelperService.prototype.window;
    /**
     * @type {?}
     * @private
     */
    ScrollHelperService.prototype.DEFAULT_SCROLLSPEED;
    /**
     * @type {?}
     * @private
     */
    ScrollHelperService.prototype.SCROLL_BUFFER;
    /**
     * @type {?}
     * @private
     */
    ScrollHelperService.prototype.document;
}
