// ng-sortgrid/esm2015/lib/mutliselect/ngsg-selection.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/mutliselect/ngsg-selection.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { fromEvent, merge, Subject } from 'rxjs';
import { filter, mapTo, withLatestFrom } from 'rxjs/operators';
import { NgsgClassService } from '../helpers/class/ngsg-class.service';
import { NgsgElementsHelper } from '../helpers/element/ngsg-elements.helper';
import { NgsgStoreService } from '../store/ngsg-store.service';
import * as i0 from "@angular/core";
import * as i1 from "../helpers/class/ngsg-class.service";
import * as i2 from "../store/ngsg-store.service";
/** @enum {number} */
const ChangeAction = {
    ADD: 0,
    REMOVE: 1,
};
ChangeAction[ChangeAction.ADD] = 'ADD';
ChangeAction[ChangeAction.REMOVE] = 'REMOVE';
/**
 * @record
 */
function SelectionChange() { }
if (false) {
    /** @type {?} */
    SelectionChange.prototype.key;
    /** @type {?} */
    SelectionChange.prototype.item;
    /** @type {?} */
    SelectionChange.prototype.action;
}
export class NgsgSelectionService {
    /**
     * @param {?} classService
     * @param {?} ngsgStore
     */
    constructor(classService, ngsgStore) {
        this.classService = classService;
        this.ngsgStore = ngsgStore;
        this.COMMAND_KEY = 'Meta';
        this.CONTROL_KEY = 'Control';
        this.selectionChange$ = new Subject();
        /** @type {?} */
        const selectionKeyPressed$ = this.selectionKeyPressed();
        this.selectionChange$
            .pipe(withLatestFrom(selectionKeyPressed$))
            .subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ([selectionChange, selectionKeyPressed]) => {
            selectionKeyPressed
                ? this.handleSelectionChange(selectionChange)
                : this.resetSelectedItems(selectionChange.key);
        }));
    }
    /**
     * @private
     * @param {?} group
     * @return {?}
     */
    resetSelectedItems(group) {
        this.ngsgStore.getSelectedItems(group).forEach((/**
         * @param {?} item
         * @return {?}
         */
        item => this.classService.removeSelectedClass(item.node)));
        this.ngsgStore.resetSelectedItems(group);
    }
    /**
     * @private
     * @param {?} selectionChange
     * @return {?}
     */
    handleSelectionChange(selectionChange) {
        if (selectionChange.action === ChangeAction.ADD) {
            this.classService.addSelectedClass(selectionChange.item);
            this.ngsgStore.addSelectedItem(selectionChange.key, {
                node: selectionChange.item,
                originalIndex: NgsgElementsHelper.findIndex(selectionChange.item)
            });
        }
        if (selectionChange.action === ChangeAction.REMOVE) {
            this.classService.removeSelectedClass(selectionChange.item);
            this.ngsgStore.removeSelectedItem(selectionChange.key, selectionChange.item);
        }
    }
    /**
     * @private
     * @return {?}
     */
    selectionKeyPressed() {
        /** @type {?} */
        const selectionKeyPressed = fromEvent(window, 'keydown').pipe(filter((/**
         * @param {?} keyboardEvent
         * @return {?}
         */
        (keyboardEvent) => keyboardEvent.key === this.COMMAND_KEY || keyboardEvent.key === this.CONTROL_KEY)), mapTo(true));
        /** @type {?} */
        const keyup = fromEvent(window, 'keyup').pipe(mapTo(false));
        return merge(selectionKeyPressed, keyup);
    }
    /**
     * @param {?} group
     * @param {?} dragedElement
     * @return {?}
     */
    selectElementIfNoSelection(group, dragedElement) {
        if (this.ngsgStore.hasSelectedItems(group)) {
            return;
        }
        this.ngsgStore.addSelectedItem(group, {
            node: dragedElement,
            originalIndex: NgsgElementsHelper.findIndex(dragedElement)
        });
    }
    /**
     * @param {?} key
     * @param {?} item
     * @param {?} selected
     * @return {?}
     */
    updateSelectedDragItem(key, item, selected) {
        this.selectionChange$.next({
            key,
            item,
            action: selected ? ChangeAction.ADD : ChangeAction.REMOVE
        });
    }
}
NgsgSelectionService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgsgSelectionService.ctorParameters = () => [
    { type: NgsgClassService },
    { type: NgsgStoreService }
];
/** @nocollapse */ NgsgSelectionService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgsgSelectionService_Factory() { return new NgsgSelectionService(i0.ɵɵinject(i1.NgsgClassService), i0.ɵɵinject(i2.NgsgStoreService)); }, token: NgsgSelectionService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgsgSelectionService.prototype.COMMAND_KEY;
    /**
     * @type {?}
     * @private
     */
    NgsgSelectionService.prototype.CONTROL_KEY;
    /**
     * @type {?}
     * @private
     */
    NgsgSelectionService.prototype.selectionChange$;
    /**
     * @type {?}
     * @private
     */
    NgsgSelectionService.prototype.classService;
    /**
     * @type {?}
     * @private
     */
    NgsgSelectionService.prototype.ngsgStore;
}
