// ng-sortgrid/esm2015/lib/helpers/class/ngsg-class.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/helpers/class/ngsg-class.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class NgsgClassService {
    constructor() {
        this.SELECTED_DEFAULT_CLASS = 'ng-sg-selected';
        this.PLACEHOLDER_DEFAULT_CLASS = 'ng-sg-placeholder';
        this.DROPPED_DEFAULT_CLASS = 'ng-sg-dropped';
    }
    /**
     * @param {?} element
     * @return {?}
     */
    addPlaceHolderClass(element) {
        element.classList.add(this.PLACEHOLDER_DEFAULT_CLASS);
    }
    /**
     * @param {?} element
     * @return {?}
     */
    removePlaceHolderClass(element) {
        element.classList.remove(this.PLACEHOLDER_DEFAULT_CLASS);
    }
    /**
     * @param {?} element
     * @return {?}
     */
    addDroppedClass(element) {
        element.classList.add(this.DROPPED_DEFAULT_CLASS);
    }
    /**
     * @param {?} element
     * @return {?}
     */
    removeDroppedClass(element) {
        element.classList.remove(this.DROPPED_DEFAULT_CLASS);
    }
    /**
     * @param {?} element
     * @return {?}
     */
    addSelectedClass(element) {
        element.classList.add(this.SELECTED_DEFAULT_CLASS);
    }
    /**
     * @param {?} element
     * @return {?}
     */
    removeSelectedClass(element) {
        element.classList.remove(this.SELECTED_DEFAULT_CLASS);
    }
}
NgsgClassService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ NgsgClassService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgsgClassService_Factory() { return new NgsgClassService(); }, token: NgsgClassService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgsgClassService.prototype.SELECTED_DEFAULT_CLASS;
    /**
     * @type {?}
     * @private
     */
    NgsgClassService.prototype.PLACEHOLDER_DEFAULT_CLASS;
    /**
     * @type {?}
     * @private
     */
    NgsgClassService.prototype.DROPPED_DEFAULT_CLASS;
}
