// ng-sortgrid/esm2015/lib/sort/sort/ngsg-sort.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sort/sort/ngsg-sort.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { timer } from 'rxjs';
import { NgsgClassService } from '../../helpers/class/ngsg-class.service';
import { NgsgElementsHelper } from '../../helpers/element/ngsg-elements.helper';
import { NgsgStoreService } from '../../store/ngsg-store.service';
import * as i0 from "@angular/core";
import * as i1 from "../../helpers/class/ngsg-class.service";
import * as i2 from "../../store/ngsg-store.service";
export class NgsgSortService {
    /**
     * @param {?} classService
     * @param {?} ngsgStore
     */
    constructor(classService, ngsgStore) {
        this.classService = classService;
        this.ngsgStore = ngsgStore;
    }
    /**
     * @param {?} group
     * @return {?}
     */
    initSort(group) {
        this.dragIndex = this.ngsgStore.getFirstSelectItem(group).originalIndex;
        this.dragElements = this.ngsgStore.getSelectedItems(group);
    }
    /**
     * @param {?} dropElement
     * @return {?}
     */
    sort(dropElement) {
        /** @type {?} */
        const hoverIndex = NgsgElementsHelper.findIndex(dropElement);
        /** @type {?} */
        const el = this.getSibling(dropElement, this.dragIndex, hoverIndex);
        if (this.isDropInSelection(el)) {
            return;
        }
        this.dragElements.forEach((/**
         * @param {?} dragElement
         * @return {?}
         */
        (dragElement) => {
            /** @type {?} */
            const insertedNode = dropElement.parentNode.insertBefore(dragElement.node, el.node);
            this.classService.addPlaceHolderClass((/** @type {?} */ (insertedNode)));
        }));
        this.dragIndex = NgsgElementsHelper.findIndex(this.dragElements[0].node);
    }
    /**
     * @return {?}
     */
    endSort() {
        this.dragElements.forEach((/**
         * @param {?} dragElement
         * @return {?}
         */
        (dragElement) => {
            this.updateDropedItem(dragElement.node);
        }));
    }
    /**
     * @private
     * @param {?} dropElement
     * @param {?} dragIndex
     * @param {?} hoverIndex
     * @return {?}
     */
    getSibling(dropElement, dragIndex, hoverIndex) {
        if (dragIndex < hoverIndex) {
            return { node: dropElement.nextSibling, originalIndex: hoverIndex + 1 };
        }
        return { node: dropElement, originalIndex: hoverIndex };
    }
    /**
     * @private
     * @param {?} dropElement
     * @return {?}
     */
    isDropInSelection(dropElement) {
        return !!this.dragElements.find((/**
         * @param {?} dragElment
         * @return {?}
         */
        (dragElment) => dragElment.node === dropElement.node));
    }
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    updateDropedItem(item) {
        this.classService.removePlaceHolderClass(item);
        this.classService.addDroppedClass(item);
        this.classService.removeSelectedClass(item);
        timer(500).subscribe((/**
         * @return {?}
         */
        () => this.classService.removeDroppedClass(item)));
    }
}
NgsgSortService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
NgsgSortService.ctorParameters = () => [
    { type: NgsgClassService },
    { type: NgsgStoreService }
];
/** @nocollapse */ NgsgSortService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgsgSortService_Factory() { return new NgsgSortService(i0.ɵɵinject(i1.NgsgClassService), i0.ɵɵinject(i2.NgsgStoreService)); }, token: NgsgSortService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgsgSortService.prototype.dragIndex;
    /**
     * @type {?}
     * @private
     */
    NgsgSortService.prototype.dragElements;
    /**
     * @type {?}
     * @private
     */
    NgsgSortService.prototype.classService;
    /**
     * @type {?}
     * @private
     */
    NgsgSortService.prototype.ngsgStore;
}
