// ng-sortgrid/esm2015/lib/store/ngsg-store.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/store/ngsg-store.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function NgsgState() { }
if (false) {
    /** @type {?} */
    NgsgState.prototype.items;
    /** @type {?} */
    NgsgState.prototype.classes;
    /** @type {?} */
    NgsgState.prototype.selectedItems;
}
export class NgsgStoreService {
    constructor() {
        this.state = new Map();
    }
    /**
     * @param {?} group
     * @param {?=} items
     * @param {?=} classes
     * @return {?}
     */
    initState(group, items = [], classes = {}) {
        this.state.set(group, { items: [...items], classes, selectedItems: [] });
    }
    /**
     * @param {?} group
     * @param {?} dragElement
     * @return {?}
     */
    addSelectedItem(group, dragElement) {
        this.state.get(group).selectedItems.push(dragElement);
    }
    /**
     * @param {?} group
     * @param {?} item
     * @return {?}
     */
    removeSelectedItem(group, item) {
        /** @type {?} */
        const updatedItems = this.state.get(group).selectedItems
            .filter((/**
         * @param {?} dragElement
         * @return {?}
         */
        (dragElement) => dragElement.node !== item));
        this.setSelectedItems(group, updatedItems);
    }
    /**
     * @param {?} group
     * @param {?} items
     * @return {?}
     */
    setItems(group, items) {
        this.state.get(group).items = [...items];
    }
    /**
     * @param {?} group
     * @return {?}
     */
    getItems(group) {
        return this.state.get(group).items;
    }
    /**
     * @param {?} group
     * @return {?}
     */
    hasItems(group) {
        return this.getItems(group).length > 0;
    }
    /**
     * @param {?} group
     * @return {?}
     */
    hasGroup(group) {
        return this.state.has(group);
    }
    /**
     * @param {?} group
     * @return {?}
     */
    getSelectedItems(group) {
        return this.state.get(group).selectedItems;
    }
    /**
     * @param {?} group
     * @param {?} selectedItems
     * @return {?}
     */
    setSelectedItems(group, selectedItems) {
        this.state.get(group).selectedItems = [...selectedItems];
    }
    /**
     * @param {?} group
     * @return {?}
     */
    getFirstSelectItem(group) {
        return this.state.get(group).selectedItems[0];
    }
    /**
     * @param {?} group
     * @return {?}
     */
    hasSelectedItems(group) {
        return this.getSelectedItems(group).length > 0;
    }
    /**
     * @param {?} group
     * @return {?}
     */
    resetSelectedItems(group) {
        this.setSelectedItems(group, []);
    }
    /**
     * @param {?} group
     * @return {?}
     */
    getClasses(group) {
        return this.state.get(group).classes;
    }
}
NgsgStoreService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ NgsgStoreService.ɵprov = i0.ɵɵdefineInjectable({ factory: function NgsgStoreService_Factory() { return new NgsgStoreService(); }, token: NgsgStoreService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgsgStoreService.prototype.state;
}
