// @angular/compiler-cli/ngcc/src/entry_point_finder/tracing_entry_point_finder.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/ngcc/src/entry_point_finder/tracing_entry_point_finder", ["require", "exports", "@angular/compiler-cli/ngcc/src/entry_point_finder/utils"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.TracingEntryPointFinder = void 0;
    var utils_1 = require("@angular/compiler-cli/ngcc/src/entry_point_finder/utils");
    /**
     * An EntryPointFinder that starts from a set of initial files and only returns entry-points that
     * are dependencies of these files.
     *
     * This is faster than processing all entry-points in the entire file-system, and is used primarily
     * by the CLI integration.
     *
     * There are two concrete implementations of this class.
     *
     * * `TargetEntryPointFinder` - is given a single entry-point as the initial entry-point. This can
     *   be used in the synchronous CLI integration where the build tool has identified an external
     *   import to one of the source files being built.
     * * `ProgramBasedEntryPointFinder` - computes the initial entry-points from the source files
     *   computed from a `tsconfig.json` file. This can be used in the asynchronous CLI integration
     *   where the `tsconfig.json` to be used to do the build is known.
     */
    var TracingEntryPointFinder = /** @class */ (function () {
        function TracingEntryPointFinder(fs, config, logger, resolver, basePath, pathMappings) {
            this.fs = fs;
            this.config = config;
            this.logger = logger;
            this.resolver = resolver;
            this.basePath = basePath;
            this.pathMappings = pathMappings;
            this.basePaths = null;
        }
        /**
         * Search for Angular package entry-points.
         */
        TracingEntryPointFinder.prototype.findEntryPoints = function () {
            var unsortedEntryPoints = new Map();
            var unprocessedPaths = this.getInitialEntryPointPaths();
            while (unprocessedPaths.length > 0) {
                var path = unprocessedPaths.shift();
                var entryPointWithDeps = this.getEntryPointWithDeps(path);
                if (entryPointWithDeps === null) {
                    continue;
                }
                unsortedEntryPoints.set(entryPointWithDeps.entryPoint.path, entryPointWithDeps);
                entryPointWithDeps.depInfo.dependencies.forEach(function (dep) {
                    if (!unsortedEntryPoints.has(dep)) {
                        unprocessedPaths.push(dep);
                    }
                });
            }
            return this.resolver.sortEntryPointsByDependency(Array.from(unsortedEntryPoints.values()));
        };
        /**
         * Parse the path-mappings to compute the base-paths that need to be considered when finding
         * entry-points.
         *
         * This processing can be time-consuming if the path-mappings are complex or extensive.
         * So the result is cached locally once computed.
         */
        TracingEntryPointFinder.prototype.getBasePaths = function () {
            if (this.basePaths === null) {
                this.basePaths = utils_1.getBasePaths(this.logger, this.basePath, this.pathMappings);
            }
            return this.basePaths;
        };
        return TracingEntryPointFinder;
    }());
    exports.TracingEntryPointFinder = TracingEntryPointFinder;
});
