// @syncfusion/ej2-spreadsheet/src/spreadsheet/integrations/number-format.js
import { refreshCellElement, rowFillHandler, getTextSpace } from '../../workbook/common/event';
import { isNullOrUndefined } from '@syncfusion/ej2-base';
import { getTextWidth } from '../common/index';
/**
 * Specifies number format.
 */
var NumberFormat = /** @class */ (function () {
    function NumberFormat(parent) {
        this.parent = parent;
        this.addEventListener();
        //Spreadsheet.Inject(WorkbookNumberFormat);
    }
    NumberFormat.prototype.refreshCellElement = function (args) {
        var cell = this.parent.getCell(args.rowIndex, args.colIndex);
        if (!isNullOrUndefined(cell)) {
            this.parent.refreshNode(cell, args);
        }
    };
    NumberFormat.prototype.getTextSpace = function (args) {
        args.width = getTextWidth(args.char, args.cell.style, this.parent.cellStyle);
    };
    NumberFormat.prototype.rowFillHandler = function (args) {
        var cellElem = args.td ? args.td : this.parent.getCell(args.rowIdx, args.colIdx);
        var span2;
        var span3;
        if (cellElem) {
            if (args.formatText) {
                cellElem.innerHTML = args.formatText.toString();
            }
            if (args.secText) {
                span3 = this.parent.createElement('span');
                span3.classList.add('e-fill-sec');
                span3.innerHTML = args.secText.toString();
            }
            if (cellElem.children.length) {
                span2 = cellElem.querySelector('.e-fill');
            }
            else {
                span2 = this.parent.createElement('span');
                span2.style.flexGrow = '1';
                span2.classList.add('e-fill');
                cellElem.appendChild(span2);
                if (span3) {
                    cellElem.appendChild(span3);
                }
                cellElem.style.display = 'flex';
            }
            span2.innerHTML = '';
            var width = getTextWidth(args.value.toString(), args.cell.style, this.parent.cellStyle);
            var count = Math.round(span2.offsetWidth / width);
            args.formatText = args.value.repeat(count);
            span2.innerHTML = args.formatText;
        }
    };
    /**
     * Adding event listener for number format.
     *
     * @hidden
     * @returns {void} - Adding event listener for number format.
     */
    NumberFormat.prototype.addEventListener = function () {
        this.parent.on(refreshCellElement, this.refreshCellElement, this);
        this.parent.on(rowFillHandler, this.rowFillHandler, this);
        this.parent.on(getTextSpace, this.getTextSpace, this);
    };
    /**
     * Removing event listener for number format.
     *
     * @hidden
     * @returns {void} - Removing event listener for number format.
     */
    NumberFormat.prototype.removeEventListener = function () {
        if (!this.parent.isDestroyed) {
            this.parent.off(refreshCellElement, this.refreshCellElement);
            this.parent.off(rowFillHandler, this.rowFillHandler);
            this.parent.off(getTextSpace, this.getTextSpace);
        }
    };
    /**
     * To Remove the event listeners.
     *
     * @returns {void} - To Remove the event listeners.
     */
    NumberFormat.prototype.destroy = function () {
        this.removeEventListener();
        this.parent = null;
    };
    /**
     * Get the workbook import module name.
     *
     * @returns {string} - Get the workbook import module name.
     */
    NumberFormat.prototype.getModuleName = function () {
        return 'numberFormat';
    };
    return NumberFormat;
}());
export { NumberFormat };
